File: spec/factories/teaching_material.rb
# frozen_string_literal: true

require 'faker'

FactoryBot.define do
  factory :teaching_material do
    lesson

    title         { Faker::Lorem.words(number: 6).join ' ' }
    subtitle      { Faker::Lorem.words(number: 6).join ' ' }
    sequence(:position)
    link          { Faker::Internet.url }
    listing_title { Faker::Lorem.words(number: 6).join ' ' }
    listing_icon  { 'paperclip' }

    image do
      file = Rails.root.join('spec', 'fixtures', 'test_image.png')
      Rack::Test::UploadedFile.new(file)
    end

    medium_type do
      medium_types = TeachingMaterial.medium_types.keys
      medium_types[rand(medium_types.length - 1)]
    end

    lesson_content do
      medium_type.to_sym == :medium_type_video
    end
  end
end

File: frontend/models/teaching-material.js
import PropTypes from "prop-types";
import { attr, fk } from "redux-orm";

import BaseModel from ".";

class TeachingMaterial extends BaseModel {
  get isVideo() {
    return this.mediumType === "medium_type_video";
  }

  get includeRef() {
    return Object.assign(super.includeRef, {
      isVideo: this.isVideo,
    });
  }
}

TeachingMaterial.modelName = "TeachingMaterial";

TeachingMaterial.fields = {
  ...BaseModel.fields,
  id: attr(),
  position: attr(),
  title: attr(),
  subtitle: attr(),
  description: attr(),
  mediumType: attr(),
  image: attr(),
  icon: attr(),
  link: attr(),
  listingTitle: attr(),
  listingIcon: attr(),
  lessonContent: attr(),
  listingItem: attr(),
  lesson: fk("Lesson", "teachingMaterials"),
};

TeachingMaterial.defaultProps = {
  ...BaseModel.defaultProps,
  id: null,
  position: null,
  title: "",
  subtitle: "",
  description: "",
  mediumType: null,
  icon: null,
  link: null,
  image: null,
  listingTitle: "",
  listingIcon: null,
  lessonContent: false,
  listingItem: false,
};

const Shape = PropTypes.shape({
  id: PropTypes.string,
  position: PropTypes.number,
  title: PropTypes.string,
  subtitle: PropTypes.string,
  description: PropTypes.string,
  mediumType: PropTypes.string,
  icon: PropTypes.string,
  link: PropTypes.string,
  image: PropTypes.string,
  listingTitle: PropTypes.string,
  listingIcon: PropTypes.string,
  lessonContent: PropTypes.bool,
  listingItem: PropTypes.bool,
});

export default TeachingMaterial;
export { Shape };
